### PR TITLE
Fix argument of invoke function being passed to OpEnqueueKernel

### PR DESCRIPTION
**Pull request by AlexeySotkin**

### CODE REVIEW

**Comment by AlexeySotkin** [comment]
Hi Sam!
Could you take a look?
Thanks,
Alexey S.

File: lib/SPIRV/SPIRVLowerOCLBlocks.cpp [hunk 1]
```diff
@@ -405,6 +405,28 @@ class SPIRVLowerOCLBlocks: public ModulePass {
       if (FirstArg->getName() != ".block_descriptor")
         continue;
 
+      bool isEnqueuedFunction = false;
```
Comment (yxsamliu) [inline lib/SPIRV/SPIRVLowerOCLBlocks.cpp]: This function is too complicated. I'd like to separate this part as a new member function

  bool isEnqueuedFunction(Function &F);

and I think more strict checking should be done to identify if a function is enqueued function:
1. F is bitcasted to a pointer P
2. P is passed to a function call C of _Z21__spirv_EnqueueKernel

and add a comment about these two conditions that are used to decide if a function is enqueued function.

Thanks.

File: lib/SPIRV/SPIRVLowerOCLBlocks.cpp [hunk 2]
```diff
@@ -405,6 +405,28 @@ class SPIRVLowerOCLBlocks: public ModulePass {
       if (FirstArg->getName() != ".block_descriptor")
         continue;
 
+      bool isEnqueuedFunction = false;
+      for (User* U : F.users()) {
+        if(!isa<ConstantExpr>(U))
+          continue;
+        // lowerBlockBind() should already have substituted last argument
+        // of enqueue_kernel() call with arguments of spir_block_bind().
+        for(User *UU : U->users())
```
Comment (yxsamliu) [inline lib/SPIRV/SPIRVLowerOCLBlocks.cpp]: space after 'for'

File: lib/SPIRV/SPIRVLowerOCLBlocks.cpp [hunk 3]
```diff
@@ -405,6 +405,28 @@ class SPIRVLowerOCLBlocks: public ModulePass {
...
+        // lowerBlockBind() should already have substituted last argument
+        // of enqueue_kernel() call with arguments of spir_block_bind().
+        for(User *UU : U->users())
+          if (CallInst* C = dyn_cast<CallInst>(UU))
+            if (Function *CF = C->getCalledFunction())
+              if (CF->getName().startswith("_Z21__spirv_EnqueueKernel")) {
+                isEnqueuedFunction = true;
+                break;
+              }
+        if(isEnqueuedFunction)
```
Comment (yxsamliu) [inline lib/SPIRV/SPIRVLowerOCLBlocks.cpp]: space

File: lib/SPIRV/SPIRVLowerOCLBlocks.cpp [hunk 4]
```diff
@@ -405,6 +405,28 @@ class SPIRVLowerOCLBlocks: public ModulePass {
...
+          break;
+      }
+      // Skip the function if it is an enqueued kernel, because per SPIR-V spec,
+      // function invoked with OpEnqueueKernel must have at least one argument
+      // of type i8*
+      if(isEnqueuedFunction)
```
Comment (yxsamliu) [inline lib/SPIRV/SPIRVLowerOCLBlocks.cpp]: space

**Comment by AlexeySotkin** [comment]
Hi Sam!
I have applied your comments. Please, take a look. Thanks.
Alexey S.

**Comment by yxsamliu** [comment]
Sorry I am on vacation. Thanks Alexey Bader for handling this.

**Pull request merged**